File: AudioVisualizer2/app/src/main/java/com/projapps/audiovisualizer/MainActivity.java
package com.projapps.audiovisualizer;

import android.media.AudioFormat;
import android.media.AudioRecord;
import android.media.MediaRecorder;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.FrameLayout;

import com.jjoe64.graphview.BarGraphView;
import com.jjoe64.graphview.GraphView;
import com.jjoe64.graphview.GraphViewSeries;
import com.jjoe64.graphview.LineGraphView;

import ca.uol.aig.fftpack.RealDoubleFFT;

public class MainActivity extends AppCompatActivity {
    private static final int SAMPLE_RATE = 44100;
    private static final int CHANNEL_CONFIG = AudioFormat.CHANNEL_IN_MONO;
    private static final int AUDIO_FORMAT = AudioFormat.ENCODING_PCM_16BIT;

    private AudioRecord mAudioRecord = null;
    private FrameLayout mLayout = null;
    private GraphViewSeries mGraphSeries = null;
    private Thread mRecordThread = null;

    private int mMinBufferSize = 0;
    private boolean isRecording = false;
    private boolean isBarGraph = true;

    private RealDoubleFFT transformer;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        mLayout = (FrameLayout) findViewById(R.id.layout);
        mMinBufferSize = AudioRecord.getMinBufferSize(SAMPLE_RATE, CHANNEL_CONFIG, AUDIO_FORMAT);
    }

    @Override
    protected void onStart() {
        super.onStart();
        initGraphView();
    }

    @Override
    protected void onPause() {
        super.onPause();
        stopRecording();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle item selection
        switch (item.getItemId()) {
            case R.id.device_access_mic:
                startRecording();
                invalidateOptionsMenu();
                return true;
            case R.id.device_access_mic_muted:
                stopRecording();
                invalidateOptionsMenu();
                return true;
            case R.id.action_settings:
                initGraphView();
                return true;
            case R.id.privacy_policy:
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public boolean onPrepareOptionsMenu(Menu menu) {
        MenuItem item1, item2;
        if (mAudioRecord == null) {
            item1 = menu.findItem(R.id.device_access_mic_muted);
            item2 = menu.findItem(R.id.device_access_mic);
        } else {
            item1 = menu.findItem(R.id.device_access_mic);
            item2 = menu.findItem(R.id.device_access_mic_muted);
        }
        if (item1 != null) {
            item1.setEnabled(false);
            item1.setVisible(false);
            if (item2 != null) {
                item2.setEnabled(true);
                item2.setVisible(true);
            }
        }
        return true;
    }

    public void startRecording() {
        mAudioRecord = new AudioRecord(MediaRecorder.AudioSource.MIC, SAMPLE_RATE, CHANNEL_CONFIG, AUDIO_FORMAT, mMinBufferSize);
        mAudioRecord.startRecording();
        isRecording = true;

        mRecordThread = new Thread(new Runnable() {
            @Override
            public void run() {
                updateGraphView();
            }
        });
        mRecordThread.start();
    }

    public void stopRecording() {
        if (mAudioRecord != null) {
            if (isRecording) {
                mAudioRecord.stop();
                isRecording = false;
                mRecordThread = null;
            }
            mAudioRecord.release();
            mAudioRecord = null;
            mGraphSeries.resetData(new GraphView.GraphViewData[]{});
        }
    }

    public void initGraphView() {
        GraphView graphView;

        if (isBarGraph) {
            graphView = new LineGraphView(this, "Time Domain");
        } else {
            graphView = new BarGraphView(this, "Frequency Domain");
        }
        isBarGraph = !isBarGraph;

        mGraphSeries = new GraphViewSeries(new GraphView.GraphViewData[]{});
        graphView.addSeries(mGraphSeries);

        if (mLayout.getChildCount() > 0) {
            mLayout.removeAllViews();
        }
        mLayout.addView(graphView);
    }

    public void updateGraphView() {
        short[] audioData = new short[mMinBufferSize];
        while (isRecording) {
            int read = mAudioRecord.read(audioData, 0, mMinBufferSize);
            if (read != AudioRecord.ERROR_INVALID_OPERATION && read != AudioRecord.ERROR_BAD_VALUE) {
                int num = audioData.length;
                final GraphView.GraphViewData[] data = new GraphView.GraphViewData[num];
                if (isBarGraph) {
                    // apply Fast Fourier Transform here
                    transformer = new RealDoubleFFT(num);
                    double[] toTransform = new double[num];
                    for (int i = 0; i < num; i++) {
                        toTransform[i] = (double) audioData[i] / Short.MAX_VALUE;
                    }
                    transformer.ft(toTransform);
                    for (int i = 0; i < num; i++) {
                        data[i] = new GraphView.GraphViewData(i, toTransform[i]);
                    }
                } else {
                    for (int i = 0; i < num; i++) {
                        data[i] = new GraphView.GraphViewData(i, audioData[i]);
                    }
                }
                MainActivity.this.runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        mGraphSeries.resetData(data);
                    }
                });
            }
        }
    }
}

File: app/src/main/java/com/projapps/audiovisualizer/MainActivity.kt
package com.projapps.audiovisualizer

import android.content.Intent
import android.media.AudioFormat
import android.media.AudioRecord
import android.media.MediaRecorder
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.Menu
import android.view.MenuItem
import ca.uol.aig.fftpack.RealDoubleFFT
import com.jjoe64.graphview.series.BarGraphSeries
import com.jjoe64.graphview.series.BaseSeries
import com.jjoe64.graphview.series.DataPoint
import com.jjoe64.graphview.series.LineGraphSeries
import kotlinx.android.synthetic.main.activity_main.*



const val SAMPLE_RATE = 44100
const val CHANNEL_CONFIG = AudioFormat.CHANNEL_IN_MONO
const val AUDIO_FORMAT = AudioFormat.ENCODING_PCM_16BIT

class MainActivity : AppCompatActivity() {

    private var mAudioRecord: AudioRecord? = null
    private var mRecordThread: Thread? = null
    private var mBaseSeries: BaseSeries<DataPoint>? = null

    private var mMinBufferSize = 0
    private var isRecording = false
    private var isBarGraph = true

    private var transformer: RealDoubleFFT? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        mMinBufferSize = AudioRecord.getMinBufferSize(SAMPLE_RATE, CHANNEL_CONFIG, AUDIO_FORMAT)
    }

    override fun onStart() {
        super.onStart()
        initGraphView()
    }

    override fun onPause() {
        super.onPause()
        stopRecording()
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.main, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        // Handle item selection
        when (item?.itemId) {
            R.id.device_access_mic -> {
                startRecording()
                invalidateOptionsMenu()
                return true
            }
            R.id.device_access_mic_muted -> {
                stopRecording()
                invalidateOptionsMenu()
                return true
            }
            R.id.action_settings -> {
                initGraphView()
                return true
            }
            R.id.privacy_policy -> {
                startActivity(Intent(this, PolicyActivity::class.java))
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onPrepareOptionsMenu(menu: Menu?): Boolean {
        val item1: MenuItem?
        val item2: MenuItem?

        if (mAudioRecord == null) {
            item1 = menu?.findItem(R.id.device_access_mic_muted)
            item2 = menu?.findItem(R.id.device_access_mic)
        } else {
            item1 = menu?.findItem(R.id.device_access_mic)
            item2 = menu?.findItem(R.id.device_access_mic_muted)
        }

        if (item1 != null) {
            item1.isEnabled = false
            item1.isVisible = false
            if (item2 != null) {
                item2.isEnabled = true
                item2.isVisible = true
            }
        }

        return super.onPrepareOptionsMenu(menu)
    }

    private fun startRecording() {
        mAudioRecord = AudioRecord(MediaRecorder.AudioSource.MIC, SAMPLE_RATE, CHANNEL_CONFIG, AUDIO_FORMAT, mMinBufferSize)
        mAudioRecord!!.startRecording()
        isRecording = true

        mRecordThread = Thread(Runnable { updateGraphView() })
        mRecordThread!!.start()
    }

    private fun stopRecording() {
        if (mAudioRecord != null) {
            if (isRecording) {
                mAudioRecord?.stop()
                isRecording = false
                mRecordThread = null
            }
            mAudioRecord?.release()
            mAudioRecord = null
            mBaseSeries?.resetData(arrayOf<DataPoint>())
        }
    }

    private fun initGraphView() {
        if (isBarGraph) {
            mBaseSeries = LineGraphSeries<DataPoint>(arrayOf<DataPoint>())
            graph.title = "Time Domain"
        } else {
            mBaseSeries = BarGraphSeries<DataPoint>(arrayOf<DataPoint>())
            graph.title = "Frequency Domain"
        }

        isBarGraph = !isBarGraph

        if (graph.series.count() > 0) {
            graph.removeAllSeries()
        }
        graph.addSeries(mBaseSeries)
    }

    private fun updateGraphView() {
        val audioData = ShortArray(mMinBufferSize)
        while (isRecording) {
            val read = mAudioRecord!!.read(audioData, 0, mMinBufferSize)
            if (read != AudioRecord.ERROR_INVALID_OPERATION && read != AudioRecord.ERROR_BAD_VALUE) {
                val num = audioData.size
                val data = arrayOfNulls<DataPoint>(num)
                if (isBarGraph) {
                    // apply Fast Fourier Transform here
                    transformer = RealDoubleFFT(num)
                    val toTransform = DoubleArray(num)
                    for (i in 0 until num) {
                        toTransform[i] = audioData[i].toDouble() / Short.MAX_VALUE
                    }
                    transformer!!.ft(toTransform)
                    for (i in 0 until num) {
                        data[i] = DataPoint(i.toDouble(), toTransform[i])
                    }
                } else {
                    for (i in 0 until num) {
                        data[i] = DataPoint(i.toDouble(), audioData[i].toDouble())
                    }
                }
                this@MainActivity.runOnUiThread { mBaseSeries!!.resetData(data) }
            }
        }
    }
}
